File: packages/python/agentguard/integrations/langchain.py
"""AgentGuard — LangChain Integration (Python).

Drop-in callback handler that evaluates every LangChain tool call through
AgentGuard's policy engine. Structurally compatible with LangChain's
BaseCallbackHandler interface — no hard dep on langchain is required.

Usage::

    from agentguard.integrations import langchain_guard

    # Pass as a callback to AgentExecutor or any LangChain chain:
    handler = langchain_guard(api_key="ag_...")

    executor = AgentExecutor.from_agent_and_tools(
        agent=agent,
        tools=tools,
        callbacks=[handler],
    )

On a block decision, ``AgentGuardBlockError`` is raised — LangChain will
surface it as a tool execution error to the caller.
"""
import json
from typing import Any, Dict, Optional, Union

from ..client import AgentGuard
from .errors import AgentGuardBlockError


class AgentGuardCallbackHandler:
    """LangChain callback handler that intercepts every tool call.

    Evaluates each tool call through the AgentGuard policy engine before
    allowing execution to proceed. Throws ``AgentGuardBlockError`` on block
    or require_approval decisions.

    Structurally compatible with ``langchain.callbacks.base.BaseCallbackHandler``
    — pass an instance anywhere LangChain accepts a callback handler.

    Args:
        api_key:   AgentGuard API key (``ag_...``)
        base_url:  Optional override for the AgentGuard API base URL
        agent_id:  Optional agent ID for scoped evaluations
    """

    name = "AgentGuardCallbackHandler"

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.agentguard.tech",
        agent_id: Optional[str] = None,
    ) -> None:
        self._agent_id = agent_id
        self._guard = AgentGuard(api_key=api_key, base_url=base_url)

    def on_tool_start(
        self,
        serialized: Dict[str, Any],
        input_str: Union[str, Dict[str, Any]],
        **kwargs: Any,
    ) -> None:
        """Called by LangChain before every tool invocation.

        Evaluates the tool call against the AgentGuard policy engine.
        Raises ``AgentGuardBlockError`` if the decision is "block" or
        "require_approval".

        Args:
            serialized:  LangChain tool serialization (must contain ``name``)
            input_str:   Tool input as a string or dict
        """
        tool_name: str = serialized.get("name", serialized.get("id", ["unknown"])[-1])

        # Normalise input
        if isinstance(input_str, str):
            try:
                parsed = json.loads(input_str)
                params: Dict[str, Any] = parsed if isinstance(parsed, dict) else {"input": input_str}
            except (json.JSONDecodeError, ValueError):
                params = {"input": input_str}
        elif isinstance(input_str, dict):
            params = input_str
        else:
            params = {"input": str(input_str)}

        result = self._guard.evaluate(tool=tool_name, params=params)

        # Normalise field names (API returns snake_case or camelCase depending on version)
        decision = result.get("result", result.get("decision", "allow"))

        if decision in ("block", "require_approval"):
            raise AgentGuardBlockError({
                **result,
                "tool": tool_name,
                "decision": decision,
                "agent_id": self._agent_id,
            })

    def on_tool_end(self, output: str, **kwargs: Any) -> None:
        """Called after successful tool execution. No-op."""

    def on_tool_error(self, error: BaseException, **kwargs: Any) -> None:
        """Called after tool execution errors. No-op."""


def langchain_guard(
    api_key: str,
    base_url: str = "https://api.agentguard.tech",
    agent_id: Optional[str] = None,
) -> AgentGuardCallbackHandler:
    """Create an AgentGuard LangChain callback handler.

    One-liner integration::

        from agentguard.integrations import langchain_guard

        executor = AgentExecutor.from_agent_and_tools(
            agent=agent,
            tools=tools,
            callbacks=[langchain_guard(api_key="ag_...")],
        )

    Args:
        api_key:   AgentGuard API key (``ag_...``)
        base_url:  Optional override for the AgentGuard API base URL
        agent_id:  Optional agent ID for scoped evaluations

    Returns:
        An ``AgentGuardCallbackHandler`` instance ready to pass as a callback.
    """
    return AgentGuardCallbackHandler(api_key=api_key, base_url=base_url, agent_id=agent_id)

File: packages/sdk/src/integrations/langchain.ts
/**
 * AgentGuard — LangChain Integration
 *
 * Drop-in callback handler that wraps every LangChain tool call through
 * AgentGuard's policy engine. Works with LangChain's BaseCallbackHandler
 * interface (structurally compatible — no hard dep on @langchain/core).
 *
 * Usage:
 *   import { langchainGuard } from '@the-bot-club/agentguard';
 *
 *   const guard = langchainGuard({ apiKey: 'ag_...' });
 *
 *   // With LangChain agent executor:
 *   const executor = AgentExecutor.fromAgentAndTools({
 *     agent,
 *     tools,
 *     callbacks: [guard],
 *   });
 *
 * On a block decision, AgentGuardBlockError is thrown — LangChain will catch
 * it and surface it to the caller as a tool execution error.
 */
import { AgentGuard } from '../sdk/client.js';
import { AgentGuardBlockError } from './errors.js';

// ─── Options ──────────────────────────────────────────────────────────────────

export interface LangChainGuardOptions {
  /** AgentGuard API key (ag_...) */
  apiKey: string;
  /** Override the AgentGuard API base URL (default: https://api.agentguard.tech) */
  baseUrl?: string;
  /** Optional agent ID to scope evaluations */
  agentId?: string;
}

// ─── LangChain-compatible tool descriptor (structural, no hard dep) ───────────

interface LangChainToolDescriptor {
  name: string;
  [key: string]: unknown;
}

// ─── AgentGuardCallbackHandler ────────────────────────────────────────────────

/**
 * LangChain callback handler that intercepts every tool call and evaluates
 * it through AgentGuard before allowing execution to proceed.
 *
 * Implements the subset of LangChain's BaseCallbackHandler that is needed
 * for tool interception. Compatible with LangChain.js 0.1.x and 0.2.x.
 *
 * Structurally compatible — does NOT require @langchain/core as a hard dep.
 * Pass an instance directly as a callback handler.
 */
export class AgentGuardCallbackHandler {
  /** Required by LangChain's callback handler interface */
  readonly name = 'AgentGuardCallbackHandler';

  private readonly guard: AgentGuard;
  private readonly options: LangChainGuardOptions;

  constructor(options: LangChainGuardOptions) {
    this.options = options;
    this.guard = new AgentGuard({
      apiKey: options.apiKey,
      baseUrl: options.baseUrl,
    });
  }

  /**
   * Called by LangChain's callback system before every tool invocation.
   *
   * Evaluates the tool call against the AgentGuard policy engine.
   * Throws AgentGuardBlockError if the decision is "block" or "require_approval".
   *
   * @param tool   The LangChain tool descriptor (must have a `name` property)
   * @param input  The serialized tool input (string or object)
   */
  async handleToolStart(
    tool: LangChainToolDescriptor,
    input: string | Record<string, unknown>,
  ): Promise<void> {
    // Normalise input — LangChain sometimes passes a JSON string
    let params: Record<string, unknown>;
    if (typeof input === 'string') {
      try {
        const parsed = JSON.parse(input) as unknown;
        params = typeof parsed === 'object' && parsed !== null
          ? (parsed as Record<string, unknown>)
          : { input };
      } catch {
        params = { input };
      }
    } else {
      params = input;
    }

    const result = await this.guard.evaluate({
      tool: tool.name,
      params,
    });

    // Map API response field names (result vs decision)
    const decision = (result as unknown as { result?: string; decision?: string }).result
      ?? (result as unknown as { decision?: string }).decision;

    if (decision === 'block' || decision === 'require_approval') {
      throw new AgentGuardBlockError({
        ...(result as unknown as Record<string, unknown>),
        tool: tool.name,
        decision,
        agentId: this.options.agentId,
      });
    }
  }

  /**
   * Called by LangChain after tool execution completes successfully.
   * No-op here — AgentGuard only intercepts before execution.
   */
   
  async handleToolEnd(_output: string): Promise<void> {
    // No-op — post-execution hook, no AgentGuard action needed
  }

  /**
   * Called by LangChain if tool execution throws an error.
   * No-op here — errors are already handled by the caller.
   */
   
  async handleToolError(_err: unknown): Promise<void> {
    // No-op
  }
}

// ─── Factory Function ─────────────────────────────────────────────────────────

/**
 * Create an AgentGuard LangChain callback handler.
 *
 * One-liner integration:
 *
 * ```typescript
 * import { langchainGuard } from '@the-bot-club/agentguard';
 *
 * const executor = AgentExecutor.fromAgentAndTools({
 *   agent,
 *   tools,
 *   callbacks: [langchainGuard({ apiKey: 'ag_...' })],
 * });
 * ```
 *
 * @param options  Guard configuration (apiKey, optional baseUrl and agentId)
 * @returns        An AgentGuardCallbackHandler instance ready to pass as a callback
 */
export function langchainGuard(options: LangChainGuardOptions): AgentGuardCallbackHandler {
  return new AgentGuardCallbackHandler(options);
}
